nvme_gw_server.py

```diff
@@ -528,74 +535,100 @@ def nvmf_subsystem_remove_host(self, request, context=None):
     def nvmf_subsystem_add_listener(self, request, context=None):
         """Adds a listener at the given TCP/IP address for the given subsystem."""
         self.logger.info({
-            f"Adding {request.trtype} listener at {request.traddr} : {request.trsvcid} for {request.nqn}"
+            f"Adding {request.gateway} {request.trtype} listener at {request.traddr} : {request.trsvcid} for {request.nqn}"
         })
 
-        try:
-            if request.trtype.lower() not in self.transports:
-                raise Exception(f"{request.trtype} transport is not enabled")
-
-            return_string = self.spdk_rpc.nvmf.nvmf_subsystem_add_listener(
-                self.spdk_rpc_client,
-                nqn=request.nqn,
-                trtype=request.trtype,
-                traddr=request.traddr,
-                trsvcid=request.trsvcid,
-                adrfam=request.adrfam,
-            )
-            self.logger.info(f"Status of add listener: {return_string}")
-        except Exception as ex:
-            self.logger.error(f"Add Listener failed: \n {ex}")
-            if context:
-                context.set_code(grpc.StatusCode.INTERNAL)
-                context.set_details(f"{ex}")
-            return pb2.req_status()
+        if not request.gateway or request.gateway == self.gateway_name:
+            try:
+                traddr = request.traddr
+                if not request.gateway:
+                    if traddr:
+                        raise Exception(f"{request.traddr} address reqires gateway")
+                    traddr = self.nvme_config.get("config", "gateway_addr")
+
+                if request.trtype.lower() not in self.transports:
```

Comment:
We could just let SPDK handle this. It would avoid this duplication of code in add and delete listener. But I don't have a strong opinion on it. So feel free to leave it if you like it better this way.

Reply:
This check was added earlier, not in this PR. Personally I like a little more this way, because the error message seems a little more clear for users. But I don't mind letting SPDK handle this. I may add a commit that removes this `self.transports` thing.